kernel/src/transaction/mod.rs

```diff
@@ -309,21 +365,95 @@ impl Transaction {
+    ) -> DeltaResult<(EngineDataResultIterator<'a>, Option<i64>)> {
+        if self.add_files_metadata.is_empty() {
+            return Ok((Box::new(iter::empty()), None));
+        }
+
+        let needs_row_tracking = self
+            .read_snapshot
+            .table_configuration()
+            .should_write_row_tracking();
+
+        let (add_actions, row_tracking_high_watermark): (
+            EngineDataResultIterator<'_>,
+            Option<i64>,
+        ) = if needs_row_tracking {
+            let row_id_high_water_mark =
+                RowTrackingDomainMetadata::get_high_water_mark(&self.read_snapshot, engine)?;
+            let mut row_tracking_visitor = RowTrackingVisitor::new(row_id_high_water_mark);
+            let mut base_row_id_batches = Vec::with_capacity(self.add_files_metadata.len());
+
+            // Visit all files to collect row tracking info
+            for add_files_batch in &self.add_files_metadata {
+                row_tracking_visitor.visit_rows_of(add_files_batch.deref())?;
+                base_row_id_batches.push(row_tracking_visitor.base_row_ids.clone());
+            }
+
+            // Create extended add files with row tracking columns
+            let extended_add_files = self.add_files_metadata.iter().zip(base_row_id_batches).map(
+                move |(add_files_batch, base_row_ids)| {
+                    let commit_versions = vec![commit_version as i64; base_row_ids.len()];
+                    let base_row_ids_array =
+                        ArrayData::try_new(ArrayType::new(DataType::LONG, true), base_row_ids)?;
+                    let commit_versions_array =
+                        ArrayData::try_new(ArrayType::new(DataType::LONG, true), commit_versions)?;
+
+                    add_files_batch.append_columns(
+                        with_row_tracking_cols(&Arc::new(StructType::new(vec![]))),
+                        vec![base_row_ids_array, commit_versions_array],
+                    )
+                },
+            );
+
+            let add_actions = self.create_add_actions(
+                engine,
+                extended_add_files,
+                with_row_tracking_cols(add_files_schema()),
+                as_log_add_schema(with_row_tracking_cols(&with_stats_col(
+                    mandatory_add_file_schema(),
+                ))),
+            );
+
+            (
+                Box::new(add_actions),
+                Some(row_tracking_visitor.row_id_high_water_mark),
+            )
+        } else {
+            // Simple case without row tracking
+            let add_actions = self.create_add_actions(
+                engine,
+                self.add_files_metadata.iter().map(|a| Ok(a.deref())),
+                add_files_schema().clone(),
+                as_log_add_schema(with_stats_col(mandatory_add_file_schema())),
+            );
+
+            (Box::new(add_actions), None)
+        };
+
+        Ok((add_actions, row_tracking_high_watermark))
```

Comment:
this split doesn't quite feel right - perhaps we should just have three things change (add files, input/output schema) and just call this once at the end? Or just inline?